soh/soh/Enhancements/randomizer/3drando/shops.cpp

```diff
@@ -130,74 +151,94 @@ int GetPriceFromMax(int max) {
+            return Random(0, 501);
+         }
+         totalWeight = totalWeight - noWeight;
+         if (totalWeight <= 0){
+            return 0;
+         }
+         totalWeight = totalWeight - childWeight;
+         if (totalWeight <= 0){
+            return Random(1, 99);
+         }
+         totalWeight = totalWeight - adultWeight;
+         if (totalWeight <= 0){
+            return Random(100, 200);
+         }
+         totalWeight = totalWeight - giantWeight;
+         if (totalWeight <= 0){
+            return Random(201, 500);
+         }
+         return Random(501, 999);
+      }
+   }
+   return 69; //this should never happen, if it does, EASTER EGG that tells us something is wrong
 }
 
-int GetRandomShopPrice() {
+uint16_t GetRandomPrice(Rando::Location *loc, PriceSettingsStruct priceSettings) {
+   uint16_t initalPrice = GetPriceFromSettings(loc, priceSettings);
    auto ctx = Rando::Context::GetInstance();
-    // If Shopsanity prices aren't Balanced, but Affordable is on, don't GetPriceFromMax
-    if (ctx->GetOption(RSK_SHOPSANITY_PRICES_AFFORDABLE).Is(true) && ctx->GetOption(RSK_SHOPSANITY_PRICES).IsNot(RO_SHOPSANITY_PRICE_BALANCED)) {
-        return GetPriceAffordable();
-    }
-
-    // max 0 means Balanced is selected, and thus shouldn't trigger GetPriceFromMax
-    int max = 0;
-
-    // check settings for a wallet tier selection and set max amount as method for setting true randomization
-    if(ctx->GetOption(RSK_SHOPSANITY_PRICES).Is(RO_SHOPSANITY_PRICE_STARTER)) {
-        max = 19; // 95/5
-    }
-    else if (ctx->GetOption(RSK_SHOPSANITY_PRICES).Is(RO_SHOPSANITY_PRICE_ADULT)) {
-        max = 40; // 200/5
-    }
-    else if (ctx->GetOption(RSK_SHOPSANITY_PRICES).Is(RO_SHOPSANITY_PRICE_GIANT)) {
-        max = 100; // 500/5
-    }
-    else if (ctx->GetOption(RSK_SHOPSANITY_PRICES).Is(RO_SHOPSANITY_PRICE_TYCOON)) {
-        max = 199; // 995/5
-    }
-    if (max != 0) {
-        return GetPriceFromMax(max);
-    }
-    // Balanced is default, so if all other known cases fail, fall back to Balanced
-    int price = 150; // JUST in case something fails with the randomization, return sane price for balanced
-    double random = RandomDouble(); //Randomly generated probability value
-    for (size_t i = 0; i < ShopPriceProbability.size(); i++) {
-        if (random < ShopPriceProbability[i]) {
-        //The randomly generated value has surpassed the total probability up to this point, so this is the generated price
-            price = i * 5; //i in range [0, 59], output in range [0, 295] in increments of 5
-            break;
-        }
-    }
-    return price;
+   if (ctx->GetOption(priceSettings.affordible) && !ctx->GetOption(priceSettings.main).Is(RO_PRICE_FIXED)){
+      if (initalPrice >= 501){
+         return 501;
+      } else if (initalPrice >= 201){
+         return 201;
+      } else if (initalPrice >= 100){
+         return 100;
+      } else if (initalPrice >= 1){
+         return 1;
+      }
```

Comment:
Maybe this would be better as it keeps the multiple of 5 prices.
```suggestion
      if (initalPrice > 500) {
         return 505;
      } else if (initalPrice > 200) {
         return 205;
      } else if (initalPrice > 99) {
         return 100;
      } else if (initalPrice > 0) {
         return 5;
      }
```